sdk/core/core/test/cmocka/test_az_url_encode.c

```diff
@@ -51,16 +54,112 @@ static void test_url_encode(void** state)
   az_span const buffer = AZ_SPAN_FROM_BUFFER(buf);
   int32_t url_length = 0;
 
-  assert_true(az_succeeded(
-      _az_span_url_encode(buffer, AZ_SPAN_FROM_STR("https://vault.azure.net"), &url_length)));
+  {
+    // Empty
+    uint8_t buf0[0] = {};
+    az_span const buffer0 = AZ_SPAN_FROM_BUFFER(buf0);
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer0, { 0 }, &url_length)));
+    assert_true(az_span_is_content_equal(az_span_slice(buffer0, 0, url_length), { 0 }));
+    assert_int_equal(url_length, 0);
+  }
+  {
+    // Just enough to do no encoding
+    uint8_t buf5[5] = { 0 };
+    az_span const buffer5 = AZ_SPAN_FROM_BUFFER(buf5);
+
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer5, AZ_SPAN_FROM_STR("hello"), &url_length)));
+
+    assert_true(
+        az_span_is_content_equal(az_span_slice(bufer5, 0, url_length), AZ_SPAN_FROM_STR("hello")));
+
+    assert_int_equal(url_length, sizeof("hello"));
+
+    // Insufficient size
+    url_length = 0xFF;
+    assert_true(
+        _az_span_url_encode(az_span_slice(0, 2), AZ_SPAN_FROM_STR("/"), &url_length)
+        == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
+
+    assert_int_equal(url_length, 0);
+    assert_true(az_span_is_content_equal(
+        az_span_slice(buffer5, 0, sizeof(buf5)), AZ_SPAN_FROM_STR("%2Flo")));
```

Comment:
This test makes it seem like we wrote past the destination slice. If the destination is only 2 bytes, this should be "%2llo", no? If we are changing the destination to "%2Flo", that is a bug.

Reply:
It should've been be either "%2Fllo", or "hello" (unmodified). We either fit a single 3-character sequence or not. I updated unit tests to include more cases like this, it should be clearer now.